File: clients/ios/Lextures/Core/Routing/LinkOpener.swift
import Foundation
import UIKit
import SwiftUI

/// Single entry point for opening URLs from app UI (MB.1 FR-1).
/// Feature code MUST call this instead of `openURL` / `UIApplication.shared.open`.
@MainActor
enum LinkOpener {
    struct Request: Equatable {
        var urlString: String
        var source: String
        /// When set, overrides shell policy (tests / forced system).
        var forceSystemBrowser: Bool = false
        /// Optional access token for first-party WebItem loads.
        var accessToken: String? = nil
        /// When true, Report appears in overflow (user-generated surfaces).
        var allowReport: Bool = false
    }

    /// Classify and act. Native destinations are handed to `shell.openDeepLink`.
    /// In-app browser is presented via `shell.presentInAppBrowser`.
    @discardableResult
    static func open(_ request: Request, shell: AppShellModel) -> MobileLinkPolicy.Classification {
        let state = policyState(from: shell)
        var classification = MobileLinkPolicy.classify(urlString: request.urlString, state: state)
        if request.forceSystemBrowser,
           classification == .inAppBrowser || classification == .systemBrowser {
            classification = .systemBrowser
        }

        switch classification {
        case .native:
            let dest = DeepLinkRouter.resolve(request.urlString)
            shell.openDeepLink(dest)
            emit(source: request.source, classification: "internal", outcome: "native")

        case .inAppBrowser:
            guard let url = resolveURL(request.urlString) else {
                emit(source: request.source, classification: "external", outcome: "error", errorClass: "bad_url")
                return .blocked
            }
            shell.presentInAppBrowser(
                InAppBrowserSession(
                    initialURL: url,
                    accessToken: request.accessToken,
                    source: request.source,
                    allowReport: request.allowReport
                )
            )
            emit(source: request.source, classification: "external", outcome: "in_app")

        case .systemBrowser:
            guard let url = resolveURL(request.urlString) else {
                emit(source: request.source, classification: "external", outcome: "error", errorClass: "bad_url")
                return .blocked
            }
            UIApplication.shared.open(url)
            emit(source: request.source, classification: "external", outcome: "system")

        case .externalApp:
            guard let url = resolveURL(request.urlString) else {
                emit(source: request.source, classification: "external", outcome: "error", errorClass: "bad_url")
                return .blocked
            }
            UIApplication.shared.open(url)
            emit(source: request.source, classification: "external", outcome: "external_app")

        case .authSession:
            // Auth flows own ASWebAuthenticationSession; if we land here from content,
            // fall back to system browser rather than capturing in WKWebView.
            if let url = resolveURL(request.urlString) {
                UIApplication.shared.open(url)
            }
            emit(source: request.source, classification: "external", outcome: "auth_session")

        case .blocked:
            shell.presentLinkBlockedNotice()
            emit(source: request.source, classification: "external", outcome: "blocked")
        }

        return classification
    }

    /// Convenience for simple string URLs.
    @discardableResult
    static func open(_ urlString: String, shell: AppShellModel, source: String) -> MobileLinkPolicy.Classification {
        open(Request(urlString: urlString, source: source), shell: shell)
    }

    /// Convenience for `URL`.
    @discardableResult
    static func open(_ url: URL, shell: AppShellModel, source: String) -> MobileLinkPolicy.Classification {
        open(Request(urlString: url.absoluteString, source: source), shell: shell)
    }

    /// OpenURLAction bridge for SwiftUI `.environment(\.openURL, …)`.
    static func openURLAction(shell: AppShellModel, source: String = "content") -> OpenURLAction {
        OpenURLAction { url in
            _ = open(url, shell: shell, source: source)
            return .handled
        }
    }

    static func policyState(from shell: AppShellModel) -> MobileLinkPolicy.State {
        MobileLinkPolicy.State(
            handling: MobileLinkPolicy.Handling.parse(shell.platformFeatures.mobileLinkHandling),
            // MB.1: in-app browser is always on; org/platform policy uses mobileLinkHandling only.
            inAppBrowserEnabled: true,
            apiHost: AppConfiguration.apiBaseURL.host ?? ""
        )
    }

    static func resolveURL(_ urlString: String) -> URL? {
        let t = urlString.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.hasPrefix("/") {
            return AppConfiguration.webURL(path: t)
        }
        return URL(string: t)
    }

    private static func emit(source: String, classification: String, outcome: String, errorClass: String? = nil) {
        var raw: [String: String] = [
            "source": source,
            "classification": classification,
            "outcome": outcome,
        ]
        if let errorClass { raw["errorClass"] = errorClass }
        let payload = MobileLinkPolicy.sanitizeTelemetry(raw)
        // Content-free counter channel; no third-party URL ever leaves the device (FR-25).
        #if DEBUG
        print("[LinkOpener]", payload)
        #endif
        InAppBrowserTelemetry.record(payload)
    }
}

/// Lightweight in-memory counters (content-free).
enum InAppBrowserTelemetry {
    private static var counts: [String: Int] = [:]

    static func record(_ payload: [String: String]) {
        let key = [
            payload["source"] ?? "unknown",
            payload["classification"] ?? "unknown",
            payload["outcome"] ?? "unknown",
        ].joined(separator: "|")
        counts[key, default: 0] += 1
    }

    static func resetForTests() { counts = [:] }
    static func snapshotForTests() -> [String: Int] { counts }
}

/// Presented in-app browser session (shell-owned).
struct InAppBrowserSession: Identifiable, Equatable {
    let id = UUID()
    var initialURL: URL
    var accessToken: String?
    var source: String
    var allowReport: Bool

    static func == (lhs: InAppBrowserSession, rhs: InAppBrowserSession) -> Bool {
        lhs.id == rhs.id
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/routing/LinkOpener.kt
package com.lextures.android.core.routing

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.lextures.android.core.config.AppConfiguration
import com.lextures.android.core.lms.MobileLinkPolicy
import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.features.home.HomeShellState

/**
 * Single entry point for opening URLs from app UI (MB.1 FR-1).
 * Feature code MUST call this instead of `Intent.ACTION_VIEW`.
 */
object LinkOpener {
    /**
     * Optional shell registered by [com.lextures.android.features.home.HomeScreen]
     * so feature call sites without an explicit shell still present the in-app browser.
     */
    @Volatile
    var activeShell: HomeShellState? = null

    data class Request(
        val urlString: String,
        val source: String,
        val forceSystemBrowser: Boolean = false,
        val accessToken: String? = null,
        val allowReport: Boolean = false,
    )

    data class BrowserSession(
        val initialUrl: String,
        val accessToken: String? = null,
        val source: String = "unknown",
        val allowReport: Boolean = false,
    )

    @JvmStatic
    fun policyState(features: MobilePlatformFeatures): MobileLinkPolicy.State =
        MobileLinkPolicy.State(
            handling = MobileLinkPolicy.Handling.parse(features.mobileLinkHandling),
            // MB.1: in-app browser is always on; org/platform policy uses mobileLinkHandling only.
            inAppBrowserEnabled = true,
            apiHost = Uri.parse(AppConfiguration.apiBaseUrl).host.orEmpty(),
        )

    @JvmStatic
    fun open(context: Context, request: Request, shell: HomeShellState?): MobileLinkPolicy.Classification {
        val resolvedShell = shell ?: activeShell
        val features = resolvedShell?.platformFeatures ?: MobilePlatformFeatures()
        val state = policyState(features)
        var classification = MobileLinkPolicy.classify(request.urlString, state)
        if (request.forceSystemBrowser &&
            (classification == MobileLinkPolicy.Classification.IN_APP_BROWSER ||
                classification == MobileLinkPolicy.Classification.SYSTEM_BROWSER)
        ) {
            classification = MobileLinkPolicy.Classification.SYSTEM_BROWSER
        }

        when (classification) {
            MobileLinkPolicy.Classification.NATIVE -> {
                val dest = DeepLinkRouter.resolve(request.urlString)
                resolvedShell?.openDeepLink(dest)
                emit(request.source, "internal", "native")
            }
            MobileLinkPolicy.Classification.IN_APP_BROWSER -> {
                val url = resolveUrl(request.urlString) ?: run {
                    emit(request.source, "external", "error", "bad_url")
                    return MobileLinkPolicy.Classification.BLOCKED
                }
                if (resolvedShell != null) {
                    resolvedShell.presentInAppBrowser(
                        BrowserSession(
                            initialUrl = url,
                            accessToken = request.accessToken,
                            source = request.source,
                            allowReport = request.allowReport,
                        ),
                    )
                    emit(request.source, "external", "in_app")
                } else {
                    // No shell available — degrade to system browser rather than drop the tap.
                    openSystem(context, url)
                    emit(request.source, "external", "system")
                }
            }
            MobileLinkPolicy.Classification.SYSTEM_BROWSER,
            MobileLinkPolicy.Classification.EXTERNAL_APP,
            MobileLinkPolicy.Classification.AUTH_SESSION,
            -> {
                val url = resolveUrl(request.urlString) ?: run {
                    emit(request.source, "external", "error", "bad_url")
                    return MobileLinkPolicy.Classification.BLOCKED
                }
                openSystem(context, url)
                val outcome = when (classification) {
                    MobileLinkPolicy.Classification.EXTERNAL_APP -> "external_app"
                    MobileLinkPolicy.Classification.AUTH_SESSION -> "auth_session"
                    else -> "system"
                }
                emit(request.source, "external", outcome)
            }
            MobileLinkPolicy.Classification.BLOCKED -> {
                resolvedShell?.presentLinkBlockedNotice()
                emit(request.source, "external", "blocked")
            }
        }
        return classification
    }

    @JvmStatic
    fun open(context: Context, urlString: String, shell: HomeShellState?, source: String): MobileLinkPolicy.Classification =
        open(context, Request(urlString = urlString, source = source), shell)

    fun resolveUrl(urlString: String): String? {
        val t = urlString.trim()
        if (t.isEmpty()) return null
        if (t.startsWith("/")) return AppConfiguration.webUrl(t)
        return runCatching { Uri.parse(t).toString() }.getOrNull()
    }

    private fun openSystem(context: Context, url: String) {
        runCatching {
            context.startActivity(
                Intent(Intent.ACTION_VIEW, Uri.parse(url)).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
            )
        }
    }

    private fun emit(source: String, classification: String, outcome: String, errorClass: String? = null) {
        val raw = buildMap {
            put("source", source)
            put("classification", classification)
            put("outcome", outcome)
            if (errorClass != null) put("errorClass", errorClass)
        }
        InAppBrowserTelemetry.record(MobileLinkPolicy.sanitizeTelemetry(raw))
    }
}

object InAppBrowserTelemetry {
    private val counts = mutableMapOf<String, Int>()

    fun record(payload: Map<String, String>) {
        val key = listOf(
            payload["source"] ?: "unknown",
            payload["classification"] ?: "unknown",
            payload["outcome"] ?: "unknown",
        ).joinToString("|")
        counts[key] = (counts[key] ?: 0) + 1
    }

    fun resetForTests() = counts.clear()
    fun snapshotForTests(): Map<String, Int> = counts.toMap()
}
